mappers/modbus/device/device.go

```diff
@@ -0,0 +1,266 @@
+}
+
+// initModbus initialize modbus client
+func initModbus(protocolConfig configmap.ModbusProtocolCommonConfig, slaveID int16, client *driver.ModbusClient) error {
+	if protocolConfig.Com.SerialPort != "" {
+		modbusRtu := driver.ModbusRtu{SlaveId: byte(slaveID),
+			SerialName:   protocolConfig.Com.SerialPort,
+			BaudRate:     int(protocolConfig.Com.BaudRate),
+			DataBits:     int(protocolConfig.Com.DataBits),
+			StopBits:     int(protocolConfig.Com.StopBits),
+			Parity:       protocolConfig.Com.Parity,
+			RS485Enabled: isRS485Enabled(protocolConfig.CustomizedValues),
+			Timeout:      5 * time.Second}
+		*client, _ = driver.NewClient(modbusRtu)
+	} else if protocolConfig.Tcp.Ip != "" {
+		modbusTcp := driver.ModbusTcp{
+			SlaveId:  byte(slaveID),
+			DeviceIp: protocolConfig.Tcp.Ip,
+			TcpPort:  strconv.FormatInt(protocolConfig.Tcp.Port, 10),
+			Timeout:  5 * time.Second}
+		*client, _ = driver.NewClient(modbusTcp)
+	} else {
+		return errors.New("No protocol found")
+	}
+	return nil
+}
+
+// initTwin initialize the timer to get twin value.
+func initTwin(dev *globals.ModbusDev) {
+	for i := 0; i < len(dev.Instance.Twins); i++ {
+		var visitorConfig configmap.ModbusVisitorConfig
+		if err := json.Unmarshal([]byte(dev.Instance.Twins[i].PVisitor.VisitorConfig), &visitorConfig); err != nil {
+			klog.Error(err)
+			continue
+		}
+		setVisitor(&visitorConfig, &dev.Instance.Twins[i], &dev.ModbusClient)
+
+		twinData := TwinData{Client: dev.ModbusClient,
+			Name:         dev.Instance.Twins[i].PropertyName,
+			Type:         dev.Instance.Twins[i].Desired.Metadatas.Type,
+			RegisterType: visitorConfig.Register,
+			Address:      visitorConfig.Offset,
+			Quantity:     uint16(visitorConfig.Limit),
+			Topic:        fmt.Sprintf(mappercommon.TopicTwinUpdate, dev.Instance.ID)}
+		collectCycle := time.Duration(dev.Instance.Twins[i].PVisitor.CollectCycle)
+		// If the collect cycle is not set, set it to 1 second.
+		if collectCycle == 0 {
+			collectCycle = 1 * time.Second
+		}
+		timer := mappercommon.Timer{twinData.Run, collectCycle, 0}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			timer.Start()
+		}()
+	}
+}
+
+// initData initialize the timer to get data.
+func initData(dev *globals.ModbusDev) {
+	for i := 0; i < len(dev.Instance.Datas.Properties); i++ {
+		var visitorConfig configmap.ModbusVisitorConfig
+		if err := json.Unmarshal([]byte(dev.Instance.Datas.Properties[i].PVisitor.VisitorConfig), &visitorConfig); err != nil {
+			klog.Error("Unmarshal visitor config failed")
+		}
+		twinData := TwinData{Client: dev.ModbusClient,
+			Name:         dev.Instance.Datas.Properties[i].PropertyName,
+			Type:         dev.Instance.Datas.Properties[i].Metadatas.Type,
+			RegisterType: visitorConfig.Register,
+			Address:      visitorConfig.Offset,
+			Quantity:     uint16(visitorConfig.Limit),
+			Topic:        dev.Instance.Datas.Topic}
+		timer := mappercommon.Timer{twinData.Run, time.Duration(dev.Instance.Twins[i].PVisitor.CollectCycle), 0}
```

Comment:
initData should same as initModbus to set a default collect cycle if dev.Instance.Twins[i].PVisitor.CollectCycle is not set ?